README.md
## Hortonworks Data Platform auto deployment by different CM

---

<img src="imgs/hortonworks_logo.png"/>

### 1. Ansible - Hortonworks Data Platform auto deployment by Ansible

### 2. Chef - Hortonworks Data Platform auto deployment by Chef

### 3. CFEngine - Hortonworks Data Platform auto deployment by CFEngine

<img src="imgs/cfengine_logo.png"/>


chef/CHANGELOG.md
# 0.1.0

Initial release of hortonworks


chef/recipes/agent.rb
include_recipe 'selinux::_common'
include_recipe 'python'
include_recipe 'python::pip'

package "epel-release" do
  action :install
end

#execute 'update system' do
##  command 'yum -y update'
#  command 'yum -y -x 'kernel*' update'
#end

bash "increase open file descriptors" do
  code <<-EOH
    ulimit -n 10000
    EOH
end

selinux_state "SELinux Disabled" do
  action :disabled
end

template "/etc/sysctl.conf" do
  path "/etc/sysctl.conf"
  source "sysctl.conf.erb"
  owner "root"
  group "root"
  mode "0644"
end

template "/etc/selinux/config" do
  path "/etc/selinux/config"
  source "selinux.erb"
  owner "root"
  group "root"
  mode "0644"
end

template "/etc/yum.repos.d/ambari.repo" do
  path "/etc/yum.repos.d/ambari.repo"
  source "ambari.repo.erb"
  owner "root"
  group "root"
  mode "0644"
end

template "/etc/hosts" do
  path "/etc/hosts"
  source "hosts.erb"
  owner "root"
  group "root"
  mode 0644
end

file "/root/.ssh/authorized_keys" do
  action :delete
end

%w{curl unzip tar wget net-tools ntp python-pip}.each  do |list|
  package list
end

service "ntpd" do
  supports :status => true, :restart => true, :reload => true
  action [ :enable, :start]
end

service "firewalld" do
  provider Chef::Provider::Service::Init::Redhat
  supports :status => true, :restart => true, :reload => true, :disable => true
  action :stop
end

directory "/root/.ssh" do
  action :create
  owner "root"
  group "root"
  mode 0755
end

template "/root/.ssh/authorized_keys" do
  path "/root/.ssh/authorized_keys"
  source "authorized_keys.erb"
  owner "root"
  group "root"
end

template "/root/.ssh/config" do
  path "/root/.ssh/config"
  source "config_ssh.erb"
  owner "root"
  group "root"
end

template "/root/.ssh/id_rsa" do
  path "/root/.ssh/id_rsa"
  source "id_rsa.erb"
  owner "root"
  group "root"
  mode 0600
end

%w{ambari-agent }.each  do |list1|
  package list1
end


template "/etc/ambari-agent/conf/ambari-agent.ini" do
  source "ambari-agent.ini.erb"
  mode 0755
  user "root"
  group "root"
end

service 'ambari-agent' do
  action [:enable, :start]
end



cfengine/masterfiles/123456/tmp/bpdeploy.sh
#!/bin/bash

REQNUM=4
ST=5
ECHOCMD="$( which echo )"
CURLCMD="$( which curl )"
GREPCMD="$( which grep )"
SLEEPCMD="$( which sleep )"
AMBSERVER="server.devops.internal"
APIURL="http://$AMBSERVER:8080/api/v1"
USPW="admin:admin"
HTAG="X-Requested-By:MyCompany"



echo $ECHOCMD
echo $CURLCMD
echo $GREPCMD
echo $SLEEPCMD


function log() {
    $ECHOCMD "[$(date +"%T")][Cluster Deployment] ${1}"
}

function wait4hosts()
{
    local HOSTNUM=$( $CURLCMD -sS -u $USPW -X GET -i $APIURL/hosts | $GREPCMD -c 'host_name' )

    while [[ $HOSTNUM -lt $REQNUM ]]
        do
            log "Required hosts number is $REQNUM and we have $HOSTNUM"
            log "Waiting until hosts number will reach $HOSTNUM, sleeping for $ST seconds"
            $SLEEPCMD $ST
        done
    $ECHOCMD "All $HOSTNUM hosts are ready for deployment"
}

#wait4hosts

# Register Blueprint with Ambari
# $CURLCMD -u $USPW -X POST -H '$HTAG' $APIURL/blueprints/AutoClusterbp?validate_topology=false -d @bp.json
# log "REGISTERING BLUEPRINT WITH $HOSTNUM HOSTS" 
# $SLEEPCMD $ST
# Create cluster
# $CURLCMD -u $USPW -X POST -H '$HTAG' $APIURL/clusters/AutoCluster1 -d @map.json
# $ECHOCMD "CREATING CLUSTER"


chef/recipes/server.rb
include_recipe 'selinux::_common'
include_recipe 'python'
include_recipe 'python::pip'

package "epel-release" do
  action :install
end

#execute 'update system' do
##  command 'yum -y update'
#  command 'yum -y -x 'kernel*' update'
#end

bash "increase open file descriptors" do
  code <<-EOH
    ulimit -n 10000
    EOH
end

selinux_state "SELinux Disabled" do
  action :disabled
end

template "/etc/sysctl.conf" do
  path "/etc/sysctl.conf"
  source "sysctl.conf.erb"
  owner "root"
  group "root"
  mode "0644"
end

template "/etc/selinux/config" do
  path "/etc/selinux/config"
  source "selinux.erb"
  owner "root"
  group "root"
  mode "0644"
end

template "/etc/yum.repos.d/ambari.repo" do
  path "/etc/yum.repos.d/ambari.repo"
  source "ambari.repo.erb"
  owner "root"
  group "root"
  mode "0644"
end

file "/root/.ssh/authorized_keys" do
  action :delete
end

%w{curl unzip tar wget net-tools ntp python-pip}.each  do |list|
  package list
end

python_pip "python-ambariclient" do
  action :install
end

service "ntpd" do
  supports :status => true, :restart => true, :reload => true
  action [ :enable, :start]
end

service "firewalld" do
  provider Chef::Provider::Service::Init::Redhat
  supports :status => true, :restart => true, :reload => true, :disable => true
  action [:stop, :disable]
end

directory "/root/.ssh" do
  action :create
  owner "root"
  group "root"
  mode 0755
end

template "/root/.ssh/authorized_keys" do
  path "/root/.ssh/authorized_keys"
  source "authorized_keys.erb"
  owner "root"
  group "root"
end

template "/root/.ssh/config" do
  path "/root/.ssh/config"
  source "config_ssh.erb"
  owner "root"
  group "root"
end

template "/root/.ssh/id_rsa" do
  path "/root/.ssh/id_rsa"
  source "id_rsa.erb"
  owner "root"
  group "root"
  mode 0600
end

template "/etc/hosts" do
  path "/etc/hosts"
  source "hosts.erb"
  owner "root"
  group "root"
  mode 0644
end

template "/tmp/bp.json" do
  path "/tmp/bp.json"
  source "bp.json.erb"
  owner "root"
  group "root"
  mode 0644
end

template "/tmp/map.json" do
  path "/tmp/map.json"
  source "map.json.erb"
  owner "root"
  group "root"
  mode 0644
end

template "/tmp/bpdeploy.sh" do
  path "/tmp/bpdeploy.sh"
  source "bpdeploy.sh.erb"
  owner "root"
  group "root"
  mode 0755
end

%w{ambari-server ambari-agent postgresql}.each  do |list1|
  package list1
end

execute 'setup ambari-server' do
  command 'ambari-server setup -s'
end

service 'postgresql' do
  action [:enable, :start]
end

r = service 'ambari-server' do
  status_command "/etc/init.d/ambari-server status | grep 'Ambari Server running'"
  action [:enable, :start]
end

service 'ambari-agent' do
  action [:enable, :start]
end

bash "run bash script" do
  code <<-EOH
    cd /tmp && ./bpdeploy.sh &
    EOH
  action :nothing
  subscribes :run, r
end	